x-pack/plugins/security_solution/server/endpoint/routes/actions/isolation.ts

```diff
@@ -93,12 +81,8 @@ export const isolationRequestHandler = function (
       });
     }
 
-    // translate any endpoint_ids into agent_ids
-    let agentIDs = req.body.agent_ids?.slice() || [];
-    if (req.body.endpoint_ids && req.body.endpoint_ids.length > 0) {
-      const newIDs = await getAgentIDsForEndpoints(req.body.endpoint_ids, context, endpointContext);
-      agentIDs = agentIDs.concat(newIDs);
-    }
+    // fetch the Agent IDs to send the commands to
+    let agentIDs = await getMetadataForEndpoints(req.body.endpoint_ids, context, endpointContext);
```

Comment:
This does not look right.
The deduce below is doing nothing because it's attempting to dedupe objects and not agent IDs. Perhaps this would be better:

```suggestion
    let agentIDs = (await getMetadataForEndpoints(req.body.endpoint_ids, context, endpointContext)).map(
      (metadata: HostMetadata) => metadata.elastic.agent.id
    );
```

With this in place, dedupe should happen and you can revert the change below on line 121.

_(also - we probably should have a test case that validates dedupe is happening)_

Reply:
I just noticed that you actually need the metadata documents that are returned, so you will need to keep those around. New suggestion:

- Rename this variable `endpointsMetadata`
- change line 86 below to:

```
// dedupe
const agentIDs = [...new Set(endpointMetadata.map((metadata: HostMetadata) => metadata.elastic.agent.id))]; 
```

Reply:
you're right that the deduping needs to be updated now that it's not just IDs. regular ol' `filter()` will be able to do it

Reply:
`endpoints.filter((current, idx, arr) => arr.map(ep => ep.elastic.agent.id).indexOf(current.elastic.agent.id) === idx)`

This _is_ O(n^2), since you're re-mapping inside the filter each time. But I can't imagine that this will be called for any `n` that is large enough to matter (thousands of agents, likely, to even take an extra millisecond of time).